0023-communication-protocol.adoc

```diff
@@ -0,0 +1,388 @@
+
+==== Sharing identities
+
+Atomic swap protocols center around the idea of providing a time window in which the change in ownership of assets is atomic.
+A core aspect of these protocols are identities.
+The communication protocol must provide a way for both parties to share identities with the other party.
+For on-chain (i.e. Layer 1) protocols like HAN and HErc20, this is always a pair of identities (redeem + refund).
+For Layer 2 protocols like HALight, the identity is a single public key: the identifier of the user's lnd instance.
+We have to accommodate for these differences in protocols.
+
+== Recommendation
+
+Based on the research above, this section presents a proposal for a set of libp2p protocols that can be combined to offer the required functionality.
+
+=== Announcing a swap
+
+Protocol-identifier: `/comit/swap/announce/1.0.0`
+
+This protocol includes two messages:
+
+. The announce message
+. The confirmation message
+
+==== Announce-message
+
+[source,json]
+----
+{
+  "swap_digest": "12202c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae"
+}
+----
+
+This message is always sent by Alice.
+See <<Assumptions and goals,assumptions point 3>> for further explanation.
+
+The `swap_digest` field is a https://github.com/multiformats/multihash[multihash-encoded] merle root that contains all details of the swap.
+It remains to be defined, how the digest is exactly computed but the idea is the following:
+
+. Take all of the swaps information _known to the COMIT node_ (alpha ledger, alpha asset, protocols, etc) and arrange it as a tree.
+. Compute the merkle root by hashing all of the nodes together
+
+The idea is that, independently of what information about the swap is given to the COMIT node (imagine different usecases, with different negotiation protocols on top of COMIT), there is a deterministic way of computing this digest.
+If both parties, independently of each other, compute the same digest, they must be knowledgeable of the same parameters.
+This should give us enough certainty to start executing the swap assuming the other party is in good will.footnote:[It is pretty much impossible to fully protected against bad will upfront.]
+Only sending a digest has several advantages over sending the actual data of the swap for comparison:
+
+- The message format is much simpler and can be reused for every swap-protocol, independent of the cryptographic protocol in use.
+- The message format is agnostic over what the nodes are actually committing to.
+
+The nodes know, which data they used to compute the digest.
+Consequently, they are also aware of what is still missing to start the swap.
+Imagine a negotiation protocol that doesn't include expiriesfootnote:[Despite the fact that they are price-relevant.].
+The nodes could:
+
+.. either default to certain expiries and include them in the digest
+.. or simply not include them in the digest.
```

Comment:
I think this is a dangerous root as it means we cannot change the default once they are fixed (as they are not part of the merkle root, they cannot be changed).

If the default as per this communication protocol become obsolete then we would end up having to advise to always specify expiries and go against such default (as they cannot be changed).

I'd recommend to include them in the digest.

Reply:
Did you read the whole paragraph? :sweat_smile: 

This section is meant to serve as an example of how the digest works and is not a proposal of whether the expiries should be included.
Actually, I specifically state that they should be included just 3 lines above our comment it says:

> Imagine a negotiation protocol that doesn't include expiries.

With a footnote:

> Despite the fact that they are price-relevant.